x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java

```diff
@@ -0,0 +1,189 @@
+            submitListener::onFailure
+        ));
+    }
+
+    private void executeSearch(SubmitAsyncSearchRequest submitRequest, String indexName, String docID,
+                               ActionListener<AsyncSearchResponse> submitListener) {
+        final Map<String, String> originHeaders = nodeClient.threadPool().getThreadContext().getHeaders();
+        final SearchRequest searchRequest = new SearchRequest(submitRequest.getSearchRequest()) {
+            @Override
+            public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> taskHeaders) {
+                AsyncSearchId searchId = new AsyncSearchId(indexName, docID, new TaskId(nodeClient.getLocalNodeId(), id));
+                return new AsyncSearchTask(id, type, action, originHeaders, taskHeaders, searchId, reduceContextSupplier);
+            }
+        };
+
+        // trigger the async search
+        final AtomicReference<Boolean> shouldStoreResult = new AtomicReference<>();
+        AsyncSearchTask task = (AsyncSearchTask) taskManager.register("transport", SearchAction.INSTANCE.name(), searchRequest);
+        SearchProgressActionListener progressListener = task.getProgressListener();
+        searchAction.execute(task, searchRequest,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    progressListener.onResponse(response);
+                    onTaskCompletion(threadPool, task, shouldStoreResult);
+                }
+
+                @Override
+                public void onFailure(Exception exc) {
+                    progressListener.onFailure(exc);
+                    onTaskCompletion(threadPool, task, shouldStoreResult);
+                }
+            }
+        );
+
+        // and get the response asynchronously
+        GetAsyncSearchAction.Request getRequest = new GetAsyncSearchAction.Request(task.getSearchId().getEncoded(),
+            submitRequest.getWaitForCompletion(), -1);
+        nodeClient.executeLocally(GetAsyncSearchAction.INSTANCE, getRequest,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(AsyncSearchResponse response) {
+                    if (response.isRunning() || submitRequest.isCleanOnCompletion() == false) {
+                        // the task is still running and the user cannot wait more so we create
+                        // an empty document for further retrieval
+                        try {
+                            store.storeInitialResponse(originHeaders, indexName, docID, response,
+                                ActionListener.wrap(() -> {
+                                    shouldStoreResult.set(true);
+                                    submitListener.onResponse(response);
+                                }));
+                        } catch (IOException exc) {
```

Comment:
should we catch all exceptions?

Reply:
+1, I pushed https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11